src/advanced_inv.cpp

```diff
@@ -1728,62 +1780,81 @@ bool advanced_inventory::query_destination( aim_location &def )
-    if( s.id == AIM_CONTAINER ) {
-        const auto cont = s.get_container( panes[src].in_vehicle() );
-        assert( cont != nullptr );
-        assert( !cont->contents.empty() );
-        assert( &cont->contents.front() == sitem.it );
-        cont->contents.erase( cont->contents.begin() );
-        rc = true;
-    } else if( sitem.area == AIM_WORN ) {
-        rc = g->u.takeoff( sitem.it );
-    } else if( sitem.from_vehicle ) {
-        rc = s.veh->remove_item( s.vstor, sitem.it );
-    }
-
-    if( !rc ) {
-        // TODO: Make this print an error if it fails
-        g->m.i_rem( s.pos, sitem.it );
+    assert( !sitem.items.empty() );
+    bool rc = true;
+
+    while(count > 0) {
+        auto &s = squares[sitem.area];
+        if( s.id == AIM_CONTAINER ) {
+            const auto cont = s.get_container( panes[src].in_vehicle() );
+            assert( cont != nullptr );
+            assert( !cont->contents.empty() );
+            assert( &cont->contents.front() == sitem.items.front() );
+            cont->contents.erase( cont->contents.begin() );
+        } else if( sitem.area == AIM_WORN ) {
+            rc &= g->u.takeoff( sitem.items.front() );
+        } else if( sitem.from_vehicle ) {
+            rc &= s.veh->remove_item( s.vstor, sitem.items.front() );
+        } else {
+            g->m.i_rem( s.pos, sitem.items.front() );
+        }
+        if(rc == false) {
+            break;
+        }
+        sitem.items.erase(sitem.items.begin());
+        --count;
     }
+    return count;
 }
 
-bool advanced_inventory::add_item( aim_location destarea, item &new_item )
+int advanced_inventory::add_item( aim_location destarea, item &new_item, int count )
 {
+    // quick bail for no count
+    if(count <= 0) {
+        return 0;
+    }
+
     assert( destarea != AIM_ALL ); // should be a specific location instead
     bool rc = true;
-    if( destarea == AIM_INVENTORY ) {
-        g->u.i_add( new_item );
-        g->u.moves -= 100;
-        return rc;
-    } else if( destarea == AIM_WORN ) {
-        return g->u.wear_item(&new_item);
-    } else {
-        advanced_inv_area &p = squares[destarea];
-        if( panes[dest].in_vehicle() ) {
-            rc = p.veh->add_item( p.vstor, new_item );
+
+    while(count > 0) {
+        if( destarea == AIM_INVENTORY ) {
+            g->u.i_add( new_item );
+            g->u.moves -= 100;
+        } else if( destarea == AIM_WORN ) {
+            rc = g->u.wear_item(&new_item);
         } else {
-            rc = g->m.add_item_or_charges( p.pos, new_item, 0 );
+            advanced_inv_area &p = squares[destarea];
+            if( panes[dest].in_vehicle() ) {
+                rc &= p.veh->add_item( p.vstor, new_item );
+            } else {
+                rc &= g->m.add_item_or_charges( p.pos, new_item, 0 );
+            }
         }
+        // show a message to why we can't add the item
+        if( rc == false  && destarea != AIM_WORN && destarea != AIM_INVENTORY) {
```

Comment:
Me thinks, you should test for `destarea` separately (for when to display a message and when not).

Currently the value of `rc` is pretty much ignored if `destarea != AIM_WORN && destarea != AIM_INVENTORY` is `false` (adding to worn or to inventory). The condition in that case becomes `( rc == false && false )`, which is always `false`, regardless of `rc`.

This means `add_item` will always return 0 when called to add to inventory or to the worn list.